Refactor Model GraphicsLayer attaching

weinig opened the pull request:
#### 7929b2727d5204bb59fa2639b182a0c152ac0a61
<pre>
Refactor Model GraphicsLayer attaching
<a href="https://bugs.webkit.org/show_bug.cgi?id=300644">https://bugs.webkit.org/show_bug.cgi?id=300644</a>

Reviewed by Mike Wyrzykowski.

Refactors how &lt;model&gt; elements attach their backends to the GraphicsLayer
tree, replacing backend specific code in RenderLayerBacking with a single
configuration point.

Now, ModelPlayer implementations implement a new `ModelPlayer::configureGraphicsLayer(...)`
function to setup the GraphicsLayer however is needed.

* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/ModelPlayerGraphicsLayerConfiguration.h: Added.
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.mm:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
* Source/WebKit/WebProcess/Model/ARKitInlinePreviewModelPlayer.h:
* Source/WebKit/WebProcess/Model/ARKitInlinePreviewModelPlayer.mm:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ios/ARKitInlinePreviewModelPlayerIOS.h:
* Source/WebKit/WebProcess/Model/ios/ARKitInlinePreviewModelPlayerIOS.mm:
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.h:
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:

Canonical link: <a href="https://commits.webkit.org/301958@main">https://commits.webkit.org/301958@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70e84c2e04262896f7eeb547c7ac6ab05e02d907

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127618 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134889 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/79175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129490 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47889 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55796 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/97143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/79175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130566 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/38299 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114312 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/77624 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/37107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32402 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78248 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108167 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137370 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54277 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41832 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105666 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54788 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110669 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105317 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26851 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50834 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54214 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60386 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/53448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/56905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/55207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -258,6 +249,11 @@ void setOpaque(bool opaque)
     return m_id;
 }
 
+void SceneKitModelPlayer::updateGraphicsLayer(GraphicsLayer& graphicsLayer, ModelPlayerGraphicsLayerUpdate&&)
```
I'm guessing this should be `DDModelPlayer::` not `SceneKitModelPlayer::`

weinig replied:
Heh. Yeah. Copy-Pasteo.

mwyrzykowski commented on Source/WebCore/Modules/model-element/ModelPlayerGraphicsLayerConfiguration.h:
```diff
@@ -0,0 +1,49 @@
+
+#include <WebCore/Color.h>
+#include <WebCore/LayoutSize.h>
+#include <WebCore/Model.h>
```
Model could be forward declared to avoid including the header

weinig replied:
The generated constructor make the include needed.

mwyrzykowski commented on Source/WebCore/Modules/model-element/ModelPlayerClient.h:
```diff
@@ -68,9 +64,9 @@ class WEBCORE_EXPORT ModelPlayerClient : public AbstractRefCountedAndCanMakeWeak
     virtual void didUpdateBoundingBox(ModelPlayer&, const FloatPoint3D&, const FloatPoint3D&) = 0;
 #endif
 
-    virtual std::optional<PlatformLayerIdentifier> modelContentsLayerID() const = 0;
```
It looks like this is still used in DragController.cpp

```
/Volumes/Data/OpenSource/Source/WebCore/page/DragController.cpp:1442:47: error: no member named 'layerID' in 'WebCore::HTMLModelElement'
 1442 |             item.modelLayerID = modelElement->layerID();
```

weinig replied:
Thanks! Though I think it is only needed on HTMLModelElement for this, so I only added it back there, not to the client interface.

mwyrzykowski commented on Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.h:
```diff
@@ -56,9 +57,8 @@ class WEBCORE_EXPORT SceneKitModelPlayer final : public ModelPlayer, public Scen
+    void configureGraphicsLayer(GraphicsLayer&, ModelPlayerGraphicsLayerConfiguration&&) override;
     void sizeDidChange(LayoutSize) override;
-    CALayer *layer() override;
-    std::optional<LayerHostingContextIdentifier> layerHostingContextIdentifier() override;
```
this appears to be used as well:
```
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/unified-sources/UnifiedSource3-nonARC.mm:3:
/Volumes/Data/OpenSource/Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:359:131: error: use of undeclared identifier 'layerHostingContextIdentifier'; did you mean 'LayerHostingContextIdentifier'?
  359 |     RELEASE_LOG(ModelElement, "%p - ModelProcessModelPlayerProxy creating remote CA layer ctxID = %" PRIu64 " id=%" PRIu64, this, layerHostingContextIdentifier().value().toUInt64(), m_id.toUInt64());
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/unified-sources/UnifiedSource3-nonARC.mm:3:
In file included from /Volumes/Data/OpenSource/Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:36:
/Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/ModelProcessModelPlayerMessages.h:39:7: note: 'LayerHostingContextIdentifier' declared here
   39 | using LayerHostingContextIdentifier = ObjectIdentifier<LayerHostingContextIdentifierType>;
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/unified-sources/UnifiedSource3-nonARC.mm:3:
/Volumes/Data/OpenSource/Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:359:131: error: use of undeclared identifier 'layerHostingContextIdentifier'; did you mean 'LayerHostingContextIdentifier'?
  359 |     RELEASE_LOG(ModelElement, "%p - ModelProcessModelPlayerProxy creating remote CA layer ctxID = %" PRIu64 " id=%" PRIu64, this, layerHostingContextIdentifier().value().toUInt64(), m_id.toUInt64());
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/unified-sources/UnifiedSource3-nonARC.mm:3:
In file included from /Volumes/Data/OpenSource/Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:36:
/Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/ModelProcessModelPlayerMessages.h:39:7: note: 'LayerHostingContextIdentifier' declared here
   39 | using LayerHostingContextIdentifier = ObjectIdentifier<LayerHostingContextIdentifierType>;
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/unified-sources/UnifiedSource3-nonARC.mm:3:
/Volumes/Data/OpenSource/Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:361:26: error: use of undeclared identifier 'layerHostingContextIdentifier'; did you mean 'LayerHostingContextIdentifier'?
  361 |     if (auto contextID = layerHostingContextIdentifier())
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/unified-sources/UnifiedSource3-nonARC.mm:3:
In file included from /Volumes/Data/OpenSource/Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:36:
/Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/ModelProcessModelPlayerMessages.h:39:7: note: 'LayerHostingContextIdentifier' declared here
   39 | using LayerHostingContextIdentifier = ObjectIdentifier<LayerHostingContextIdentifierType>;
```

mwyrzykowski commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
```diff
@@ -218,9 +220,23 @@ void ModelProcessModelPlayer::sizeDidChange(WebCore::LayoutSize size)
+    if (!layerHostingContextIdentifier)
+        return;
+
+    m_graphicsLayer->setContentsToModelContext(
```
`m_graphicsLayer` appears to be not defined
```
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/unified-sources/UnifiedSource127.cpp:7:
/Volumes/Data/OpenSource/Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:230:5: error: use of undeclared identifier 'm_graphicsLayer'; did you mean 'graphicsLayer'?
  230 |     m_graphicsLayer->setContentsToModelContext(
```

mwyrzykowski commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
```diff
@@ -39,6 +40,7 @@
 #include <WebCore/MIMETypeRegistry.h>
 #include <WebCore/Model.h>
```
I think `#include <WebCore/ModelContext.h>` is needed for `ModelContextDisablePortal`

mwyrzykowski replied:
and `<WebCore/GraphicsLayer.h>` include too

mwyrzykowski commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
```diff
@@ -218,9 +220,23 @@ void ModelProcessModelPlayer::sizeDidChange(WebCore::LayoutSize size)
+            *layerHostingContextIdentifier,
+            graphicsLayer.primaryLayerID(),
+            configuration.contentSize,
+            configuration.hasPortal = WebCore::ModelContextDisablePortal::No : WebCore::ModelContextDisablePortal::Yes,
```
I think this assignment should probably be a `?` ternary operation

mwyrzykowski commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
```diff
@@ -218,9 +220,23 @@ void ModelProcessModelPlayer::sizeDidChange(WebCore::LayoutSize size)
+
+    m_graphicsLayer->setContentsToModelContext(
+        WebCore::ModelContext::create(
+            *layerHostingContextIdentifier,
```
I see this error compiling locally:
```
In file included from /Volumes/Data/OpenSource/WebKitBuild/Release-xros/DerivedSources/WebKit/unified-sources/UnifiedSource127.cpp:7:
/Volumes/Data/OpenSource/Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:234:13: error: no viable conversion from 'value_type' (aka 'WTF::ObjectIdentifierGeneric<WebCore::LayerHostingContextIdentifierType, WTF::ObjectIdentifierMainThreadAccessTraits<uint64_t>, unsigned long long>') to 'const PlatformLayerIdentifier' (aka 'const ProcessQualified<ObjectIdentifierGeneric<WebCore::PlatformLayerIdentifierType, ObjectIdentifierMainThreadAccessTraits<unsigned long long>, unsigned long long>>')
  234 |             *layerHostingContextIdentifier,
```

weinig committed:
Refactor Model GraphicsLayer attaching
https://bugs.webkit.org/show_bug.cgi?id=300644

Reviewed by Mike Wyrzykowski.

Refactors how <model> elements attach their backends to the GraphicsLayer
tree, replacing backend specific code in RenderLayerBacking with a single
configuration point.

Now, ModelPlayer implementations implement a new `ModelPlayer::configureGraphicsLayer(...)`
function to setup the GraphicsLayer however is needed.

* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/ModelPlayerGraphicsLayerConfiguration.h: Added.
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelPlayer.mm:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
* Source/WebKit/WebProcess/Model/ARKitInlinePreviewModelPlayer.h:
* Source/WebKit/WebProcess/Model/ARKitInlinePreviewModelPlayer.mm:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ios/ARKitInlinePreviewModelPlayerIOS.h:
* Source/WebKit/WebProcess/Model/ios/ARKitInlinePreviewModelPlayerIOS.mm:
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.h:
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:

Canonical link: https://commits.webkit.org/301958@main
